notebooks/template.ipynb

```diff
@@ -0,0 +1,494 @@
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "user:\n",
+      "  name: {{ first }} {{ last }}\n",
+      "  favorite_food: {{ food }}\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat fixtures/template/render-template.yaml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d335aec6-5fcd-4df2-ae2e-8afa1a510683",
+   "metadata": {},
+   "source": [
+    "The `values_needed` parameter can be used to display which values are needed to complete the template. A logger needs to be initialized for the log of the missing values to be displayed."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "040eceb1-0821-4e82-825a-5be18f06397d",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "[2024-08-18T09:52:37]     INFO Value(s) needed to render this template are:\n",
+      "[2024-08-18T09:52:37]     INFO   first\n",
+      "[2024-08-18T09:52:37]     INFO   food\n",
+      "[2024-08-18T09:52:37]     INFO   last\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "user:\n",
+      "  name: {{ first }} {{ last }}\n",
+      "  favorite_food: {{ food }}\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "import uwtools.logging\n",
+    "uwtools.logging.setup_logging(verbose=False)\n",
+    "\n",
+    "print(template.render(input_file='fixtures/template/render-template.yaml', values_needed=True))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6d6913b9-a375-447a-b729-566afe84f694",
+   "metadata": {},
+   "source": [
+    "The log tells us that values will be needed for keys `first`, `food`, and `last`. These values can be sourced from a Python dictionary or from a file. The following file provides the values we need:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "7f794c66-8840-419a-adf5-20efddb85708",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "first: John\n",
+      "last: Doe\n",
+      "food: burritos\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat fixtures/template/render-values.yaml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9c809a20-b09d-438a-a1da-3cb9986ce9fa",
+   "metadata": {},
+   "source": [
+    "With these values, we can render the template into a new file with all of the values included. The paths to files containing templates or values can be specified either as a string or a <a href=\"https://docs.python.org/3/library/pathlib.html#pathlib.Path\">Path</a> object.  If a values file is used to add values to the template, its file type can be specified by `values_format` to a string matching the file type. The resulting rendered file can be specified by setting `output_file` parameter. If `output_file` is not specified or set to `None`, it will be written to `stdout`. "
```

Comment:
```suggestion
    "With these values, we can render the template to a file. When the values source is a file, its path can be given either as a string or a <a href=\"https://docs.python.org/3/library/pathlib.html#pathlib.Path\">Path</a> object. If it has an unrecognized (or no) extension, its format can be specified with `values_format`. The rendered template can be written to a file specified with `output_file`; otherwise, it will be written to `stdout`. "
```
- "with all of the values included" is implied by "render".
- The file may or may not be new -- an old file will be overwritten by this operation.
- I tried to simplify a little.
- The case where the extension is unrecognized might make a good example on its own: We could show how `uwtools` complains if e.g. the filename is `render-values.txt`, then show how, if we specify `values_format='yaml'` the error disappears.

I just tried the bad-extension idea in this notebook, and see that the error output when the values-source file's extension is unrecognized is pretty, probably unnecessarily, verbose. We should talk about maybe making that a little gentler -- could you bring it up in standup or group if I forget to?